src/governance.ts

```diff
@@ -133,6 +147,51 @@ export class GovernanceCanister {
     return this.responseConverters.toListProposalsResponse(rawResponse);
   };
 
+  public stakeNeuron = async ({
+    stake,
+    principal,
+    ledgerCanister,
+  }: {
+    stake: ICP;
+    principal: Principal;
+    ledgerCanister: LedgerCanister;
+  }): Promise<NeuronId | TransferError> => {
```

Comment:
Thanks for the work on introducing strongly-typed errors here :)

When a function returns an error type, all the errors within that error type should be expected and handled by the caller. In this case, the caller of `stakeNeuron` can expect any of the errors in `TransferError` to happen.

With that in mind, using and adding new error types of `TransferError` as is done in this PR is problematic. `ledgerCanister.transfer` returns a `TransferError`, but now it has errors that have nothing to do with transferring ICP (e.g. `NeuronNotFound`), and it doesn't make sense for a client to handle a `NeuronNotFound` error when they are trying to transfer ICP.

To get around this, `stakeNeuron` and other methods should have their own error type. Something like this (not syntactically correct), and feel free to rename things as you see fit:

```
StakeNeuronError {
  // The stake is < 1 ICP
  LessThanOneICP,

  // The transfer to stake the neuron failed.
  // Here we wrap the original `TransferError` that was returned.
  TransferError(TransferError);

  // An error occurred when claiming the neuron
  ClaimNeuronFailed,
}
```

Reply:
Ah, good point, I thought `TransferError` was general, not Ledger specific. I understood `Transfer` as request, not as ledger transfers.

Do you propose to group errors per canister or methods? What do you think makes more sense?

What do you mean by `TransferError(TransferError)`? `StakeNeuronError` can return a TransferError from ledger, right?